### DISCUSSION TITLE
Routing with a direct-to-client delivery option?

### DISCUSSION THREAD

**Discussion by khanhtranadi** [PyVRP/PyVRP#717] [Q&A] [answered]
Hi team.
I have a situation that cannot be solved by the current pyvrp version (0.10).
That is the situation where the goods are sent by express delivery service. With small packages and long distances, it is impossible to use one means of transport to deliver it to the customer.

Currently, I only see the cost calculated by: Fixed, by distance, by time.
But with express types, it will be calculated by weight/volume x distance.

## Example

I have 3 trucks, capacity is [10,15,20] tons, and cost per 10km are [$2, $3, $4]

Demand from customers:
- customer 1: 8 tons
- customer 2: 0.1 tons

Distances:
0 -> 1: 50km
1 -> 2: 520km
0 -> 2: 500km

Express Service Cost: $10/10km/tons

If have no Express Service: The optimized Route is: Truck 1 from 0 -> 1 -> 2 -> 0. The total cost is  $214.

With Express Service: 
The optimized Route is: 
1) Truck 1 from 0 -> 1 -> 0. The cost is $20.
2) Express from 0-> 2. The cost is $50.
TOTAL $70 only.

**Comment by leonlan** [comment]
Hi @khanhtranadi! What exactly do you mean by costs calculated as "weight x distance"? Can you give an example?

**Reply by khanhtranadi** [reply]
Sorry for the abbreviation.
The full calculation is Unit price x weight/volume x distance.
For example:
The unit price is $0.08/ton/km and $0.06/m3/km.
So the cost of transporting 2 packages of 1.5 tons(2m3) over a distance of 350km is:
Calculated by weight: $0.08 x (2 x 1.5) x 350 = $84
Calculated by volume: $0.06 x (2x 2) x 350 = $112
The final cost is the max of the above 2 results, which is $112

**Reply by leonlan** [reply]
This is indeed not possible in PyVRP right now. It could be implemented by introducing a parameter `unit_load_distance_cost: int | list[int]` which adds an additional contribution to the costs based on each load dimension. But this use case is a bit too niche to make this part of PyVRP. @N-Wouda what do you think?

**Reply by N-Wouda** [reply]
Hi @khanhtranadi, I would like to clarify something because I am not completely sure about the cost structure yet. Suppose we have a route `0 -> 1 -> 2 -> 0` (0 is the depot).

Suppose we have the following data:
- distances: 
  - 0 -> 1: 1km
  - 1 -> 2: 2km
  - 2 -> 0: 4km
- weights and volumes: 
  - customer 1: [6 tonnes, 8 m3]
  - customer 2: [4 tonnes, 4 m3]
- unit prices:
  - $0.08/tonne/km
  - $0.06/m3/km

could you then show me how you calculate the cost value associated with the route?

**Reply by khanhtranadi** [reply]
> Hi @khanhtranadi, I would like to clarify something because I am not completely sure about the cost structure yet. Suppose we have a route `0 -> 1 -> 2 -> 0` (0 is the depot).
> 
> Suppose we have the following data:
> 
> * distances:
>   
>   * 0 -> 1: 1km
>   * 1 -> 2: 2km
>   * 2 -> 0: 4km
> * weights and volumes:
>   
>   * customer 1: [6 tonnes, 8 m3]
>   * customer 2: [4 tonnes, 4 m3]
> * unit prices:
>   
>   * $0.08/tonne/km
>   * $0.06/m3/km
> 
> could you then show me how you calculate the cost value associated with the route?

With the express service, have only one route: From depot to customer.
So there are 2 routes: 0 -> 1 and 0 -> 2.

**Reply by N-Wouda** [reply]
Where does the optimisation come in?

**Reply by khanhtranadi** [reply]
I have this example:
I have 3 trucks, capacity is [10,15,20] tons, and cost per 10km are [$2, $3, $4]

Demand from customers:
- customer 1: 8 tons
- customer 2: 0.1 tons

Distances:
0 -> 1: 50km
1 -> 2: 520km
0 -> 2: 500km

Express Service Cost: $10/10km/tons

If have no Express Service: The optimized Route is: Truck 1 from 0 -> 1 -> 2 -> 0. The total cost is  $214.

With Express Service: 
The optimized Route is: 
1) Truck 1 from 0 -> 1 -> 0. The cost is $20.
2) Express from 0-> 2. The cost is $50.
TOTAL $70 only.

So that is the optimization.

**Reply by N-Wouda** [reply, marked as answer]
Pre-compute the express cost for each client. Set these as the prizes, and mark the clients optional. See [here](https://pyvrp.org/examples/quick_tutorial.html#Prize-collecting-VRP) for details. Doing this minimises `<routing costs> + <uncollected prizes>`, which is equivalent to `<own routing> + <express costs>`. After optimising, any clients not in the solution have their goods delivered via express.

**Reply by khanhtranadi** [reply]
Good treatment plan, but I don't know how to minimizes ` <routing costs> + <uncollected prizes>.`

**Reply by N-Wouda** [reply]
PyVRP does so automatically.

**Reply by khanhtranadi** [reply]
I have read the pyvrp documentation but still can't find the code to minimizes <routing costs> + <uncollected prizes>. Can you give me a sample code?

**Reply by N-Wouda** [reply]
@khanhtranadi it's in the example I linked to above. [here](https://pyvrp.org/examples/quick_tutorial.html#Prize-collecting-VRP).

**Reply by khanhtranadi** [reply]
Yeah. Finally, I understood my problem and how prized PyVRP works. Thank @N-Wouda , @leonlan
